docs/source/extending_pywr/extending_pywr_parameters.rst

```diff
@@ -0,0 +1,224 @@
+            super().__init__(self, model)
+            self.total_volume = total_volume
+
+        def setup(self):
+            # allocate an array to hold the parameter state
+            num_scenarios = len(self.model.scenarios.combinations)
+            self._volume_remaining = np.empty([num_scenarios], np.float64)
+
+        def reset(self):
+            # reset the amount remaining in all states to the initial value
+            self._volume_remaining[...] = self.total_volume
+
+        def value(self, timestep, scenario_index):
+            # return the current volume remaining for the scenario
+            return self._volume_remaining[scenario_index.global_id]
+
+        def after(self):
+            # update the state
+            timestep = self.model.timestepper.current  # get current timestep
+            flow_during_timestep = self._node.flow * timestep.days  # array of flows
+            self._remaining -= flow_during_timestep
+            self._remaining[self._remaining < 0] = 0  # volume remaining cannot be less than zero
+
+.. [*] The model is said to be "dirty" if nodes or edges are added or removed, resulting in a change to the structure
+       of the linear programme used to solve the model. This usually requires Parameters which track state to
+       reallocate memory, instead of just resetting values.
+
+Dependency on other parameters
+==============================
+
+The value of each parameter is calculated at the start of every timestep. A dependency tree is used to ensure that
+parameters are evaluated in the correct order. For example, the ``AggregatedParameter`` returns the aggregated value
+of a set of parameters using a user-defined function. In the terminology of the dependency tree the
+``AggregatedParameter`` is the parent of the other parameters, which are it's children. When writing a parameter these
+dependencies need to be defined explicitly by modifying the ``Parameter.parents`` or ``Parameter.children`` attributes.
+
+To get the value of a child parameter use the ``Parameter.get_value`` method, or for the index use
+``Parameter.get_index``. These methods return the value/index for the current timestep and scenario. To access the
+value from previous timesteps you must manually track the state of the child parameters.
+
+The ``pywr.parameters.load_parameter`` function is used to load parameters from JSON. This works with both references to
+parameters and nested parameters.
+
+As an example, see a simplified version of ``AggregatedParameter`` that returns the sum value of it's child parameters.
+
+.. code-block:: python
+
+    class SumParameter(Parameter):
+        def __init__(self, model, parameters):
+            self.parameters = parameters
+            for parameter in self.parameters:
+                self.children.add(parameter)
+
+        def value(self, timestep, scenario_index):
+            total_value = sum([parameter.get_value() for parameter in parameters])
```

Comment:
`get_value()` takes `scenario_index` as an argument.

Reply:
Can you tell I haven't actually run any of these except the cython example? :)